Google.Api.Gax.Grpc.Tests/Rest/PartialDecodingStreamReaderTest.cs

```diff
@@ -0,0 +1,165 @@
+        /// Test data split by characters.
+        /// </summary>
+        [Fact]
+        public async void DecodingStreamReaderTestByChar()
+        {
+            
+            StreamReader reader = new MyStreamReader(DataStr.Select(c =>  c.ToString()));
+            var decodingReader = new PartialDecodingStreamReader<JObject>(Task.FromResult(reader), JObject.Parse);
+
+            var result = await decodingReader.MoveNext(CancellationToken.None).ConfigureAwait(false);
+            Assert.True(result);
+            Assert.NotNull(decodingReader.Current);
+            Assert.Equal(decodingReader.Current["foo"], 1);
+
+            result = await decodingReader.MoveNext(CancellationToken.None).ConfigureAwait(false);
+            Assert.True(result);
+            Assert.NotNull(decodingReader.Current);
+            Assert.Equal(decodingReader.Current["bar"], 2);
+
+            result = await decodingReader.MoveNext(CancellationToken.None).ConfigureAwait(false);
+            Assert.False(result);
+
+            result = await decodingReader.MoveNext(CancellationToken.None).ConfigureAwait(false);
+            Assert.False(result);
+        }
+
+        /// <summary>
+        /// Test when data breaks off unexpectedly.
+        /// </summary>
+        [Fact]
+        public async void DecodingStreamReaderTestIncomplete()
+        {
+            var s = DataStr.Split(',')[0]; // right after the first element ends
+            StreamReader reader = new MyStreamReader(s.Split('\n'));
+            var decodingReader = new PartialDecodingStreamReader<JObject>(Task.FromResult(reader), JObject.Parse);
+
+            var result = await decodingReader.MoveNext(CancellationToken.None).ConfigureAwait(false);
+            Assert.True(result);
+            Assert.NotNull(decodingReader.Current);
+            Assert.Equal(decodingReader.Current["foo"], 1);
+
+            var ex = await Assert.ThrowsAsync<InvalidOperationException>(async () =>
+                await decodingReader.MoveNext(CancellationToken.None).ConfigureAwait(false));
+            Assert.Contains("Closing `]` bracket not received after iterating through the stream.", ex.Message);
+        }
+
+        /// <summary>
+        /// Test when data is empty array.
+        /// </summary>
+        [Fact]
+        public async void DecodingStreamReaderTestEmpty()
+        {
+            StreamReader reader = new MyStreamReader(new[] {"[]"});
+            var decodingReader = new PartialDecodingStreamReader<JObject>(Task.FromResult(reader), JObject.Parse);
+
+            var result = await decodingReader.MoveNext(CancellationToken.None).ConfigureAwait(false);
+            Assert.False(result);
+
+            result = await decodingReader.MoveNext(CancellationToken.None).ConfigureAwait(false);
+            Assert.False(result);
+        }
+    }
+
+    /// <summary>
+    /// A fake of a StreamReader emitting given strings
+    /// </summary>
+    internal class MyStreamReader : StreamReader
```

Comment:
Please rename this to something more descriptive - `ReplayingStreamReader` would probably be reasonable.